cogs/poll.py

```diff
@@ -0,0 +1,318 @@
+        )    # if single choice is implemented, don't forget to correct this
+        embed.set_author(
+            name="{} created a poll with {} choices!".format(
+                username, len(choices)),
+            icon_url=pfp)
+        # eventually add this: embed.set_footer(text="5 users voted • Current winner: D (or top n if specified) •
+        # Options: Non-anonymous, Multiple choice")
+
+        # Dividing the options into pages. A new page is created when there is more than 7 options OR 550 characters
+        # (including the question) on a page. However, a page always contains at least the question and one option
+        # even if they use more than 550 characters.
+        options_list = [[]]
+        page_number = 1
+        options_on_page = 0
+        char_on_page = len(question)
+        for arg in choices:
+            if options_on_page == 0 or (char_on_page + len(arg) < 550
+                                        and options_on_page < 7):
+                options_list[page_number - 1].append(arg)
+                options_on_page += 1
+                char_on_page += len(arg)
+            else:
+                page_number += 1
+                options_list.append([])
+                options_list[page_number - 1].append(arg)
+                options_on_page = 1
+                char_on_page = len(question) + len(arg)
+
+        embed.set_footer(
+            text=
+            "Initializing... Please wait, options will soon appear (Try 1/3)"
+            .format(len(options_list)))
+        msg = await ctx.send(embed=embed)
+
+        # add the emojis to the message
+        failed = False
+        tries = 0
+        while msg:
+            try:
+                pos = 0
+                for arg in choices:
+                    await msg.add_reaction(EMOJI_ALPHABET[pos])
+                    pos += 1
+                if len(options_list) > 1:
+                    await msg.add_reaction('◀')
+                    await msg.add_reaction('▶')
+                break
+            except discord.errors.Forbidden:
+                tries += 1
+                if tries == 3:
+                    failed = True
+                    break
+                else:
+                    embed.set_footer(
+                        text=
+                        "Was unable to put all reactions on message. Please don't add any emoji "
+                        "until this text disappears. (Try {}/3)".format(tries +
+                                                                        1))
+                    await msg.edit(embed=embed)
+                    await msg.clear_reactions()
+
+        # paginate
+        p = Pages(ctx, msg, options_list, embed, timeout, timeout_unit, failed)
+
+        await p.paginate()
+        return
+        # need to also eventually add options for anonymous voting in DMs
+        # restricting to only one choice only and
+        # add a duration for polls.
+
+
+class Pages:

```

Comment:
If this class behaves similarly to Pages inside `paginator.py`, better put them in the same file and give them different names or make one class inherit from the other, or modify the original class to allow longer time out. This way you don't need to write much more code, and other cogs can benefit from it as well

Reply:
The problem is there's a couple major differences. An important one is that this one takes a message that was already sent and edits it. A main reason for that is that the reacts for the poll options must already be placed before the paginator starts. I am not sure this would be useful for other stuff than polling. Another significant difference is that it takes as input a 2D array where the items are already divided: the rows are the different pages and the columns are the items on each page (The original paginator divides the items itself). Finally, unlike the original paginator, it only edits the fields, so that the embed attributes (author, etc) aren't changed.

Reply:
If I shouldn't let it in poll.py, then I think the best option is for me to make another class in paginator.py considering all these differences, but I'm gonna need to generalize it before I can do that, because right now there's a lot of poll-specific things in it. Also, if I do that, there's things I won't be able to ever implement, like tracking what the current option with the most votes is, because it would need to be in paginator

Reply:
Counting votes is simple as long as you have the original message id. Adding emojis for options before adding the arrows can be implemented in the original paginator.

Reply:
I've decided that modifying paginator.py was the best way, adding a display_option that takes a list of embeds, an option to replace the message by a string or an embed at the end instead of deleting, and a way to input a list of emojis to put before or after the arrows. I'll commit it soon, but for now I've tested it on my machine and it works well.

Reply:
Doesn't an option for embed formatting already exist? You can modify it without needing to add new attributes. I don't think it's going to contrast any user of paginator